src/hotspot/share/utilities/growableArray.hpp

```diff
@@ -209,14 +209,16 @@ class GrowableArrayView : public GrowableArrayBase {
     return -1;
   }
 
-  int  find(void* token, bool f(void*, E)) const {
+  template<typename T>
+  int find(T* token, bool f(T*, const E&)) const {
```

Comment:
What is the advantage of a const reference here?

Reply:
You can bind a non-const reference to a const one but not the other way.

Reply:
Sorry I was unclear: what is the advantage of a reference here? Is it just to avoid copying

Reply:
@dholmes-ora Yes it helps avoid copying, especially if the copy constructor is non-trivial. And I think it is more idiomatic in C++ to use references here.

Reply:
Using a reference here leads to unnecessary overhead when `E` is small and
trivially copyable, unless the predicate function is inlined. Pass by value is
better in that case. Of course, as noted above, if `E` is "expensive" to copy
or non-copyable then a reference is needed. Boost has this thing called
`call_traits<E>::param_type` for this issue, but I don't recommend we copy
that.

Idiomatic C++ makes the entire function a template parameter, as was suggested
earlier in this PR. That dodges this question entirely, leaving the parameter
passing decision to the predicate function where it belongs, rather than
having it imposed by GrowableArray::find.  The find function just imposes the
requirement that the predicate satisfies the appropriate constraints, e.g. it
is callable on an element reference and returns convertible to bool.

Reply:
I agree we should be using a template-typed function instead of a function pointer here.
I think a lot of our uses of function pointers in our code base would work better as template-typed args.
See for example the `grow` argument (of template type `GFN`) at this point:
https://github.com/openjdk/jdk/blob/d051f22284e7ccc288c658588f73da672d9bfacd/src/hotspot/share/utilities/unsigned5.hpp#L343C34-L343C34

In that case I cited, if the `grow` argument it were a function pointer instead of a template-typed function-like argument (Kim what’s the right term here? “functor”??), then performance and flexibility would be unacceptably harmed.

I think the `find` argument is just the same kind of thingy.

Reply:
"Functor" is a reasonable choice, with plenty of precedent. 

The C++ standard doesn't use the term "functor". Instead it defines (C++14
20.9) the terms "function object type" and associated "function object" to
include all of

(1) pointer to function types
(2) class types with a member operator()
(3) class types with a conversion to pointer to function

The C++ standard doesn't have distinct terminology for cases (2) and (3), so
far as I can tell.

Sometimes people use either "functor" or "function object class" to
distinguish (2) (web search for "C++ functor" for examples of the former).
Sometimes people use "function object" to refer only to (2). So the community
is inconsistent in this respect.

I'm not sure I've ever seen (3).  (Not even in "look at this obscure C++
thing" discussions.)